Source/WebKit/WebProcess/WebPage/Cocoa/UnifiedTextReplacementController.mm

```diff
@@ -669,45 +542,126 @@ static void replaceContentsInRange(WebCore::Document& document, const WebCore::S
     if (m_contextRanges.isEmpty())
         return;
 
-    if (!m_webPage) {
+    RefPtr document = this->document();
+    if (!document) {
         ASSERT_NOT_REACHED();
         return;
     }
 
+    auto selectionRange = document->selection().selection().firstRange();
+    if (!selectionRange)
+        return;
+
+    if (!document->selection().isCaret())
+        return;
+
+    auto nodeAndMarker = findReplacementMarkerContainingRange(*document, *selectionRange);
+    if (!nodeAndMarker)
+        return;
+
+    auto& [node, marker] = *nodeAndMarker;
+    auto data = std::get<WebCore::DocumentMarker::UnifiedTextReplacementData>(marker.data());
+
+    m_webPage->textReplacementSessionUpdateStateForReplacementWithUUID(data.sessionUUID, WebTextReplacementData::State::Active, data.uuid);
+}
+
+// MARK: Private instance helper methods.
+
+std::optional<WebCore::SimpleRange> UnifiedTextReplacementController::contextRangeForSessionWithUUID(const WTF::UUID& uuid) const
+{
+    RefPtr liveRange = m_contextRanges.get(uuid);
+    return WebCore::makeSimpleRange(liveRange);
+}
+
+RefPtr<WebCore::Document> UnifiedTextReplacementController::document() const
+{
+    if (!m_webPage) {
+        ASSERT_NOT_REACHED();
+        return nullptr;
+    }
+
     RefPtr corePage = m_webPage->corePage();
     if (!corePage) {
         ASSERT_NOT_REACHED();
-        return;
+        return nullptr;
     }
 
     RefPtr frame = corePage->checkedFocusController()->focusedOrMainFrame();
     if (!frame) {
         ASSERT_NOT_REACHED();
-        return;
+        return nullptr;
     }
 
-    RefPtr document = frame->document();
+    return frame->document();
+}
+
+// This overload is only meant to be used by the other overloads of this method.
+void UnifiedTextReplacementController::replaceContentsOfRangeInSession(const WTF::UUID& uuid, const WebCore::SimpleRange& range, WTF::Function<void(WebCore::Editor&)>&& replacementOperation)
```

Comment:
I think the way we generally express this is `fooInternal()` (called from `foo()`).